### ISSUE TITLE
Can't select text for more than 1 line

### ISSUE THREAD

**Issue by CavalcanteLeo** [Dashibase/lotion#13]

**Comment by vvidday** [comment]
Hi, unfortunately I think this issue is quite complicated... The base reason for not being able to select multiple lines is due to the blocks having `contenteditable=true`. This can be circumvented by wrapping the parent element with `contenteditable=true`.

However, this raised some other problems -
1) It seems like the parent div having `contenteditable=true` "blocks" all listeners in child components, e.g. pressing enter no longer creates a new block, deleting no longer triggers a merge, because the listeners in `Block` never get executed
2) Tiptap editor becomes buggy, e.g. inserting bold/italic text moves the cursor to the front

One potential solution is to dynamically set the `contenteditable` property to only be true when the user is making a selection. This seems to work but there might be caveats that I'm not taking into account. However, even assuming that we can make it so that a user can make multiple line selections, it creates a bigger issue with regard to how we are going to deal with these selections. (e.g. deleting multiple lines)

To my knowledge, the way the project is currently set up is that individual `Blocks` are "responsible" for managing their state and data, however a deletion of multiple lines would require multiple blocks being changed based on one action.

[Here's](https://github.com/vvidday/lotion/commit/4df82377e537b593f7e8d707645f45a4e04134a3) my rough attempt at dealing with these actions. The general idea is to "lift" the logic for multiple line events to the `Lotion` component, where we use the user's selection (from `window.getSelection()`) to traverse the DOM and determine the appropriate changes to the data. However, this approach seems rather hacky and is likely to be very error prone, especially since we're kind of "taking over" handling events from the default browser implementation. 

The above code is deployed [here](https://phenomenal-crumble-c54318.netlify.app/) - it's not finished and is still buggy for 'enter', breaks on non-text blocks and doesn't handle bold or italic text.

I would greatly appreciate input on the above approach, or if anyone has a better way to approach the issue!

**Comment by CavalcanteLeo** [comment]
have you seen this [editor.js](https://editorjs.io/)?

they have a similar approach

**Comment by DAlperin** [comment]
Yeah, I have taken several cracks at this kind of problem before and there is no great answer. I think The way notion does it is the technical best, but also by far the hardest way. As far as I can tell, notion renders all the blocks into one big contenteditable with many nested contenteditables inside? Then instead of making a new "editor" per "block" the render/editing engine keeps track of the data model and seamlessly displays it as one document.

**Comment by DAlperin** [comment]
Exactly, it's all "one" editor. Even though the data model is blocks it gets rendered as a single logical editor.

**Comment by CavalcanteLeo** [comment]
multiple contenteditable <img width="929" alt="Screen Shot 2022-08-01 at 13 48 37" src="https://user-images.githubusercontent.com/1735944/182201434-67c0fba4-77e5-46c4-8ac2-6406e23dd708.png">

**Comment by DAlperin** [comment]
![image](https://user-images.githubusercontent.com/16063713/182202210-4c3a9dac-7381-4b91-bcac-8bdc2e7491c0.png)
With a parent contenteditable though

**Comment by greentfrapp** [comment]
Hey everyone, thanks for the active discussion on this topic! We've also done a little bit of investigation into this and while having a parent contenteditable might work on Chrome, it still breaks on the latest version of Firefox. Interestingly, Notion does not use a parent contenteditable on Firefox, but we haven't dug into how it's implemented yet.

**Comment by DAlperin** [comment]
@greentfrapp if you ever want to chat, I'd love to pick your brain/share some of my notes on the topic from the last time I faced this issue a little while ago. I come from the React side of things, so I don't have any good insight into the implementation side of things with Vue, but I do have some abstract knowledge built up that might be useful.

**Comment by greentfrapp** [comment]
@DAlperin I'll love to take you up on that offer! Could you drop an email to sk@dashibase.com and we'll set up some time for a video call?

**Comment by johnpuddephatt** [comment]
> One potential solution is to dynamically set the contenteditable property to only be true when the user is making a selection. This seems to work but there might be caveats that I'm not taking into account.

I've just taken a look at how Wordpress/Gutenberg handles this and this is the approach they take. In both Chrome and Firefox, I can see the contenteditable attribute being toggled on the parent element. 

[I've created a simple test case without Lotion](https://codepen.io/johnpuddephatt/pen/ZEoEJGW) and it looks like it's working fine in Chrome/Firefox/Safari.

**Comment by yuguaa** [comment]
In recent days, I have been using contenteditable to write similar requirements, but my requirements include the requirement to use the mouse to select, right-click to merge two blocks, so that they are displayed in one piece, similar to seamless text, but at the same time to support A single block clicks and triggers a response event, which makes me miserable

**Comment by yuguaa** [comment]
> In recent days, I have been using contenteditable to write similar requirements, but my requirements include the requirement to use the mouse to select, right-click to merge two blocks, so that they are displayed in one piece, similar to seamless text, but at the same time to support A single block clicks and triggers a response event, which makes me miserable

Slatejs might work, let me try it

**Comment by aigcprompt** [comment]
according to https://www.blocknotejs.org/

**Comment by nsnl-coder** [comment]
```
import { useEffect } from 'react';

const useToggleContentEditable = (
  editorRef: React.RefObject<HTMLDivElement>,
) => {
  useEffect(() => {
    if (!editorRef.current) return;

    const onMouseDown = () => {
      const selection = window.getSelection();
      if (!selection) return;

      selection.removeAllRanges();
      editorRef.current!.contentEditable = 'true';
    };

    const onMouseUp = () => {
      const selection = window.getSelection();
      if (!selection) return;

      const range = selection.getRangeAt(0);

      editorRef.current!.contentEditable = 'false';

      selection.removeAllRanges();
      selection.addRange(range);
    };

    editorRef.current.addEventListener('mousedown', onMouseDown);
    editorRef.current.addEventListener('mouseup', onMouseUp);

    return () => {
      editorRef.current?.removeEventListener('mousedown', onMouseDown);
      editorRef.current?.removeEventListener('mouseup', onMouseUp);
    };
  }, []);
};

export default useToggleContentEditable;

```
  
 Hi guys,
I try to implement @johnpuddephatt solution but when i click on child Content Editable element, it losts the focus.
Here is my implementation to restore the cursor position when user click on child content editable element using Reactjs
